torchtext/models/roberta/model.py

```diff
@@ -94,26 +98,34 @@ def forward(self, features):
 
 
 class RobertaModel(Module):
-    def __init__(self, encoder: Module, head: Optional[Module] = None):
+    """
+
+    Example - Instantiate model with user-specified configuration
+        >>> from torchtext.models import RobertaEncoderConf, RobertaModel, RobertaClassificationHead
+        >>> roberta_encoder_conf = RobertaEncoderConf(vocab_size=250002)
+        >>> encoder = RobertaModel(config=roberta_encoder_conf)
+        >>> classifier_head = RobertaClassificationHead(num_classes=2, input_dim=768)
+        >>> classifier = RobertaModel(config=roberta_encoder_conf, head=classifier_head)
+    """
+    def __init__(self, config: RobertaEncoderConf, head: Optional[Module] = None, freeze_encoder: bool = False):
```

Comment:
Do we want to continue to accept python objects in `__init__` as we were doing before and instead deal with instantiation from a config in a separate `from_config` method?

Reply:
I was thinking about the different between config-based / object-based approach here.
I do not have a recommendation, as neither is clear winner, but I came to think that the major difference is that accepting only config means that the code is allowing only a set of encoders parameterized by the config, while accepting an object means that the code is allowing any encoder with the same interface (duck-typing).

You might have already thought through, but I think that's something you want to be aware when making the choice here, so that it's aligned with the goal of this API.

Reply:
> Do we want to continue to accept python objects in `__init__` as we were doing before and instead deal with instantiation from a config in a separate `from_config` method?

I think passing python object would mean that user would need to instantiate them separately and use this class only to stitch encoder with head. Also as noted in previous comment, we might want to keep encoder as private and only expose Model as the only public facing API as this is sufficient to instantiate both only encoder or encoder+head.

Reply:
> I was thinking about the different between config-based / object-based approach here. I do not have a recommendation, as neither is clear winner, but I came to think that the major difference is that accepting only config means that the code is allowing only a set of encoders parameterized by the config, while accepting an object means that the code is allowing any encoder with the same interface (duck-typing).
> 

I think passing the dataclass config might allow for better type checking (http://blog.ezyang.com/2020/10/idiomatic-algebraic-data-types-in-python-with-dataclasses-and-union/). 

Also as noted in other comment, user would need to first instantiate separately the encoder and then make another call to Model Construction when otherwise they could have just done with one call to the constructor of RobertaModel.


> You might have already thought through, but I think that's something you want to be aware when making the choice here, so that it's aligned with the goal of this API.

Reply:
> while accepting an object means that the code is allowing any encoder with the same interface (duck-typing).


BTW, we might not want to do this, since this class is specific to RobertaEncoder. Not sure if I miss-understood your comment though? 
 
In general my mental model behind the (Roberta)Model class API is as follows:
1) Provide a simple mechanism to specify the configuration of the encoder. 
 
* I chose the dataclass config to completely define the architecture. The advantage of specifying architecture through dataclass is that it can be easily carried around in multiple places and use-sites(including inside various bundler and pipeline APIs).  
 
* Alternative would be to provide the information explicitly as constructor arguments but this would lead to duplication of architecture parameters  across multiple places and changes in encoder would inevitably require making changes in model class (or other places wherever encoder is used).
 
2) Optionally provide a head module to be used for specific tasks.
 
3) Optionally freeze the encoder to allow for fine-tuning only the head with a pre-trained encoder.

Reply:
My comment was more of my thinking out loud for API design around model construction. I see many different ways to approach like (`from_pretrained`, object-based, config-based etc...) and I was trying to see if there is a winner, and while there is no single winner, I noticed that there is one essential difference in config-based (strongly-typed) construction vs object-based construction (duck-typing).

It seems like you have a good principles for the choice of the API, and the points you listed make sense. I do not think I am miss understood. Thanks for the reply.